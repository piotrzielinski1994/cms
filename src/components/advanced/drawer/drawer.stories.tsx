import { Button } from '@/components/basic/button/button';
import type { Meta, StoryObj } from '@storybook/react';
import { useState, type ComponentProps } from 'react';
import { Drawer as DrawerComponent } from './drawer';

type Args = ComponentProps<typeof DrawerComponent>;

const meta = {
  component: DrawerComponent,
  title: 'Advanced/Drawer',
  argTypes: {},
  args: {},
} satisfies Meta<Args>;

const Render = ({ ...args }: Args) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div className="flex gap-1">
      <Button onClick={() => setIsOpen((prev) => !prev)}>Open</Button>
      <DrawerComponent {...args} isOpen={isOpen} onClose={() => setIsOpen(false)}>
        @@@
      </DrawerComponent>
    </div>
  );
};

const Drawer: StoryObj<typeof DrawerComponent> = { render: Render };

export { Drawer };
export default meta;
