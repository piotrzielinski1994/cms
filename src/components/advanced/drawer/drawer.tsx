import { HtmlProps } from '@/utils/html/html.types';
import { cn } from '@/utils/tailwind';
import { useEffect, useRef } from 'react';

type DrawerProps = HtmlProps<'div'> & {
  isOpen: boolean;
  onClose: () => void;
};

const Drawer = ({ isOpen, onClose, children, ...props }: DrawerProps) => {
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (isOpen) ref.current?.focus();
  }, [isOpen]);

  return (
    <div
      {...props}
      ref={ref}
      tabIndex={-1}
      onKeyDown={(e) => {
        if (e.key !== 'Escape') return;
        e.stopPropagation();
        onClose();
      }}
      className={cn(
        'fixed right-0 inset-y-0 z-50',
        'w-64 p-4',
        'bg-foreground text-background shadow-lg',
        `transform translate-x-full transition-transform`,
        { 'translate-x-0': isOpen },
      )}
    >
      {children}
    </div>
  );
};

export { Drawer };
