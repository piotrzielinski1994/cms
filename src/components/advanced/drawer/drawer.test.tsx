import { render } from '@testing-library/react';
import { ComponentProps } from 'react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';
import { Drawer } from './drawer';

describe('Drawer', () => {
  const defaultProps = {
    isOpen: true,
    onClose: vi.fn(),
  } satisfies ComponentProps<typeof Drawer>;

  it('should have no accessibility violations', async () => {
    const { container } = render(<Drawer {...defaultProps} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(<Drawer {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
