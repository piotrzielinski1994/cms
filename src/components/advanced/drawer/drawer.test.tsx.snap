// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Drawer > should match the snapshot 1`] = `
<div>
  <div
    class="fixed inset-y-0 right-0 w-64 bg-white shadow-lg z-50 p-4 transition-transform transform outline-none translate-x-0"
    tabindex="-1"
  />
</div>
`;
